crucible-wasm/src/Lang/Crucible/Wasm/Translate.hs

```diff
@@ -0,0 +1,861 @@
+  ]
+
+type WasmGenerator s ret = Generator WasmExt s WasmGenState ret IO
+
+data WasmGenState s =
+  WasmGenState
+  { genLocals :: Seq.Seq (Some (Reg s))
+     -- ^ Function-local variables
+  , genStack  :: Seq.Seq (StackVal s)
+     -- ^ Value stack.  Back/right end of the sequence is
+     --   the working end of the stack to match the
+     --   order presented in the Wasm specification.
+  }
+
+emptyGenState :: WasmGenState s
+emptyGenState =
+  WasmGenState
+  { genLocals = mempty
+  , genStack  = mempty
+  }
+
+addLocal :: Reg s tp -> WasmGenState s -> WasmGenState s
+addLocal r st = st{ genLocals = genLocals st Seq.|> Some r }
+
+data StackVal s
+  = StackI32 (Atom s (BVType 32))
+  | StackI64 (Atom s (BVType 64))
+  | StackF32 (Atom s (FloatType SingleFloat))
+  | StackF64 (Atom s (FloatType DoubleFloat))
+  | StackBool (Atom s BoolType)
+
+class PushStack s x where
+  pushStack :: x -> WasmGenerator s ret ()
+
+instance PushStack s (StackVal s) where
+  pushStack x = modify (\st -> st{ genStack = genStack st Seq.|> x })
+
+instance PushStack s (Atom s BoolType) where
+  pushStack = pushStack . StackBool
+
+instance PushStack s (Atom s (BVType 32)) where
+  pushStack = pushStack . StackI32
+
+instance PushStack s (Atom s (BVType 64)) where
+  pushStack = pushStack . StackI64
+
+instance PushStack s (Atom s (FloatType SingleFloat)) where
+  pushStack = pushStack . StackF32
+
+instance PushStack s (Atom s (FloatType DoubleFloat)) where
+  pushStack = pushStack . StackF64
+
+instance PushStack s (Atom s tp) => PushStack s (Expr WasmExt s tp) where
+  pushStack x = pushStack =<< mkAtom x
+
+instance PushStack s Word32 where
+  pushStack x = pushStack =<< mkAtom (App (BVLit (knownNat @32) (BV.word32 x)))
+
+instance PushStack s Word64 where
+  pushStack x = pushStack =<< mkAtom (App (BVLit (knownNat @64) (BV.word64 x)))
+
+instance PushStack s Float where
+  pushStack x = pushStack =<< mkAtom (App (FloatLit x))
+
+instance PushStack s Double where
+  pushStack x = pushStack =<< mkAtom (App (DoubleLit x))
+
+instance PushStack s ConstantValue where
+  pushStack (I32Val x) = pushStack x
+  pushStack (I64Val x) = pushStack x
+  pushStack (F32Val x) = pushStack x
+  pushStack (F64Val x) = pushStack x
+
+
+iteVals ::
+  Expr WasmExt s BoolType ->
+  StackVal s ->
+  StackVal s ->
+  WasmGenerator s ret (StackVal s)
+
+-- promote boolean values as required
+iteVals c (StackI32 x) (StackBool b) =
+  do y <- mkAtom (App (BoolToBV knownNat (AtomExpr b)))
+     iteVals c (StackI32 x) (StackI32 y)
+iteVals c (StackI64 x) (StackBool b) =
+  do y <- mkAtom (App (BoolToBV knownNat (AtomExpr b)))
+     iteVals c (StackI64 x) (StackI64 y)
+iteVals c (StackBool b) (StackI32 y) =
+  do x <- mkAtom (App (BoolToBV knownNat (AtomExpr b)))
+     iteVals c (StackI32 x) (StackI32 y)
+iteVals c (StackBool b) (StackI64 y) =
+  do x <- mkAtom (App (BoolToBV knownNat (AtomExpr b)))
+     iteVals c (StackI64 x) (StackI64 y)
```

Comment:
Should there be  promotion from I32 to I64 (and F32 to F64)?

Reply:
No, this promotion isn't actually part of the semantics, it's just a way to avoid certain noisy round-tripping operations that encode bools into bitvectors and project them back.  There's no automatic promotion in the Wasm spec.

Reply:
Gotcha.  Can you add that to the comments for this block of functions?